celery/app/task.py

```diff
@@ -882,8 +882,13 @@ def replace(self, sig):
             )
 
         if self.request.chain:
+            options = sig.options.copy()
             for t in reversed(self.request.chain):
                 sig |= signature(t, app=self.app)
+            sig.set(
+                link=options.get('link', None),
```

Comment:
What happens if we already have a `link` or `link_error` for the rest of this chain? Will we overwrite it, and would that be surprising?

Reply:
I honestly have no idea.

Reply:
Would you like to try to come up with such a test case?

Reply:
Sure thing. Sorry, this was on my list but I hadn't gotten around to it over the weekend. I'll see what I can do after work today now that you've pushed that fix for the test which raced the callback.

Reply:
I have a new test case for this and have explained what I worked out in the main thread. Will resolve this and replace with a comment suggesting the diff I'm working on now